### PR TITLE
Add hook implementation function name to plugin call order sorter

### PR DESCRIPTION
**Pull request by DragaDoncila**
# Description
As discussed at the dev meeting, having multiple hook implementations in a single file works out of the box, but each implementation is displayed by plugin name in the call order sorter, which is confusing.
plugin_call_order_original

The two display options which we agreed were most reasonable were
1. Display both the function name and the plugin name in the widget
2. Display the plugin name if the hook implementation doesn't have a custom name, and the function name if it does, leaving the plugin name to a tooltip.

Screenshots of what each option would look like:
plugin_call_order_both plugin_call_order_tooltip

I think option 1 is the nicest, but interested to hear what people would prefer.

### CODE REVIEW

File: napari/_qt/widgets/qt_plugin_sorter.py [hunk 1]
```diff
@@ -77,7 +77,25 @@ def __init__(self, item: QListWidgetItem, parent: QWidget = None):
         self.position_label = QLabel()
         self.update_position_label()
 
+        # ~~~show plugin name + function name
+        self.setToolTip("Click and drag to change call order")
         self.plugin_name_label = QLabel(item.hook_implementation.plugin_name)
+        plugin_name_font = QFont()
```
Comment (DragaDoncila) [inline napari/_qt/widgets/qt_plugin_sorter.py]: I don't think I should create a separate font for this and hardcode the font size, but I'm not sure how to manipulate the theme to do it for me.

Comment (tlambert03) [reply napari/_qt/widgets/qt_plugin_sorter.py]: one option is to use `setObjectName` on a specific object, and then add an entry to one of the qss files like `02_custom.qss` that matches that name.  For instance, lower down in this file, I used:

```python
        title = QLabel('Plugin Sorter')
        title.setObjectName("h3")
```

and then in `02_custom.qss`:

```css
QLabel#h3 {
  font-size: 18px;
  color: {{ secondary }};
}
```

It's fine if you need to do a "one-off" style declaration, but if you can use something that's already being used, all the better (such as `#small_text`)

**Comment by sofroniewn** [comment]
> I think option 1 is the nicest, but interested to hear what people would prefer.

I also agree, I really like the use of a smaller font. We can continue to iterate and standardize this stuff across napari.

Not necessarily for this PR, but as a followup we now might want to rename our builtin to something like `magic_imread` which is the function name. Not required, but probably nice to do. Curious what others think?

**Comment by tlambert03** [comment]
I too like option 1, very nice.  One additional thing you might want to do here is to use an `ElidingLabel`, which will truncate the name and add an ellipses if the window size is not big enough (otherwise, it could get very annoying with longer names).  There is an `ElidingLabel` class in `qt_eliding_label.py`:

```python
In [1]: from napari._qt.widgets.qt_eliding_label import ElidingLabel

In [2]: l = ElidingLabel("some-very-long-plugin-name")

In [3]: l.show()
```

Screen Recording 2021-02-12 at 11 53 57 AM

**Comment by DragaDoncila** [comment]
> I too like option 1, very nice. One additional thing you might want to do here is to use an `ElidingLabel`, which will truncate the name and add an ellipses if the window size is not big enough (otherwise, it could get very annoying with longer names). There is an `ElidingLabel` class in `qt_eliding_label.py`:
> 
> ```python
> In [1]: from napari._qt.widgets.qt_eliding_label import ElidingLabel
> 
> In [2]: l = ElidingLabel("some-very-long-plugin-name")
> 
> In [3]: l.show()
> ```
> 

@tlambert03 I tried changing the label to an eliding label, and it mostly works. I just can't seem to get rid of the margin on the right. If I align the plugin label right or left, then it shows ellipses no matter how wide the dialog is, even with the `MinimumExpanding` horizontal size policy. What's the best way to achieve this? Should I be adding a sub-layout?

plugin_call_order_margin

**Comment by tlambert03** [comment]
>  I just can't seem to get rid of the margin on the right.

I think the issue here is that ElidingLabel [uses `self.font()`](https://github.com/napari/napari/blob/5ebb3cc914038519591d7186a8a2d42b6fdc42b9/napari/_qt/widgets/qt_eliding_label.py#L12) to determine how "wide" the displayed text is.  But [style-sheets take precedence](https://doc.qt.io/qt-5/qwidget.html#font-prop) in the case of a font-size conflict... so it thinks text is actually larger than it is, and does the eliding incorrectly.  You might need to set the fontSize manually on this one, and not use the stylesheet.

**Comment by DragaDoncila** [comment]
> ou might need to set the fontSize manually on this one, and not use the stylesheet.

If I set a fixed font point size of 8, I get one letter of the plugin name and then ellipses regardless of setting a size policy or not. If I set it to 6 the label is not displayed at all.

plugin_call_order_fixed_font

**Comment by tlambert03** [comment]
the ElidingLabel stores the fontMetric during the `__init__`... so if you're changing font size after instantiating, it won't work.  You can change the eliding label to recalculate the fontmetric each time like this: 
```python
class ElidingLabel(QLabel):
    """A single-line eliding QLabel."""

    def __init__(self, text='', parent=None):
        super().__init__(parent)
        self._text = text

    def setText(self, txt):
        self._text = txt
        fm = QFontMetrics(self.font())
        short = fm.elidedText(self._text, Qt.ElideRight, self.width())
        super().setText(short)

    def resizeEvent(self, rEvent):
        width = rEvent.size().width()
        fm = QFontMetrics(self.font())
        short = fm.elidedText(self._text, Qt.ElideRight, width)
        super().setText(short)
        rEvent.accept()

```

**Comment by DragaDoncila** [comment]
> You can change the eliding label to recalculate the fontmetric each time

Okay this kind of works, as in it works for one hook spec. I've made the changes to eliding label and it works well for `napari_get_reader` but not for other hook specs. I tried it with fixed point size and using the `small_text` style. I thought maybe it was just the initial size of the dialog but resizing the window doesn't change the label.

plugin_call_order_reader 
plugin_call_order_writer

**Comment by tlambert03** [comment]
I'm not sure I'm following the problem.  but you may need to adjust the size policy and/or the maximum width of that label as well.  setting a different background color on one or more of the widgets can help to identify what the limitations are (e.g. if it's the label size itself that's a problem, or padding, or the eliding, or the size policy, etc...).  feel free to ping me if you want more specific tips.

**Comment by DragaDoncila** [comment]
Ok @tlambert03 thanks for all your help on this. I think I've narrowed it down to it looking mostly good, with one issue. The label expands appropriately with the window now, and shrinks appropriately too. However, the plugin name doesn't always align, depending on the length of the function name.
image

If I try to align the plugin name to the right, then the label elides and won't expand, regardless of the fact that its horizontal stretch factor is larger than the function name's label. I think this is because the stretch happens to the right, and the label is right aligned so it doesn't perceive itself as having any space. 

image

Is there a way to align the labels to the right and still have them stretch on expansion?

**Comment by tlambert03** [comment]
> Is there a way to align the labels to the right and still have them stretch on expansion?

ah interesting... is this code pushed?  (i.e. can I pull and test it out myself?)

**Comment by DragaDoncila** [comment]
Just pushed it then with the alignment and bg color

File: napari/_qt/widgets/qt_eliding_label.py [hunk 2]
```diff
@@ -9,19 +9,23 @@ class ElidingLabel(QLabel):
...
 
     def resizeEvent(self, rEvent):
         width = rEvent.size().width()
-        short = self.fm.elidedText(self._text, Qt.ElideRight, width)
+        short = self._getShortText(width)
         super().setText(short)
         rEvent.accept()
 
+    def _getShortText(self, width):
+        self.fm = QFontMetrics(self.font())
```
Comment (Carreau) [inline napari/_qt/widgets/qt_eliding_label.py:25]: oOut of curiosity why was this moved out of init / in which condition does it become stale ?

**Review by Carreau — APPROVED** [review]

File: napari/_qt/widgets/qt_eliding_label.py (hunk 2 above)
Comment (tlambert03) [reply napari/_qt/widgets/qt_eliding_label.py:25]: yeah, see https://github.com/napari/napari/pull/2255#issuecomment-779215124 ... she found that if she changed the font size after creating the label, the metric was incorrect

**Comment by sofroniewn** [comment]
> I think I've narrowed it down to it looking mostly good, with one issue. The label expands appropriately with the window now, and shrinks appropriately too. However, the plugin name doesn't always align, depending on the length of the function name.

If you can't fix this part, I think it still looks pretty good. Definitely good enough for a first pass for me personally (top screenshot in comment above that is).

**Comment by goanpeca** [comment]
>Ok @tlambert03 thanks for all your help on this. I think I've narrowed it down to it looking mostly good, with one issue. The label expands appropriately with the window now, and shrinks appropriately too. However, the plugin name doesn't always align, depending on the length of the function name.

It looks pretty good as it is @DragaDoncila 🥳 !

Now to be able to align them we would need to have some refresh mechanism to get the maximum length of the function name in all the list items and set this to be the minimum width of the other items. This would need to be triggered on resize events as well. As @sofroniewn points out, it is not critical, but doable :)

**Comment by DragaDoncila** [comment]
Alright @sofroniewn @goanpeca I've removed the debug blue highlighting and alignment. If we're happy with this as a first pass then I think this is ready to merge.

**Review by sofroniewn — APPROVED** [review]
Looks great @DragaDoncila thanks very much!!
